Implicit sumtypes don't work with channels

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.4 75830f1
**OS:** macOS

**What did you do?**
```v
struct Aa {}
struct Ab {}

type As = Aa | Ab

fn main() {
    a := chan As{}
    a <- Aa{}
    // a <- As(Aa{}) works
}
```


**What did you expect to see?**

No cgen error.

 
**What did you see instead?**

Cgen error:
```v
/tmp/v_501/test2.15413596262970579007.tmp.c:2818:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/test2.15413596262970579007.tmp.c:14349:29: error: passing 'main__Aa' (aka 'struct main__Aa') to parameter of incompatible type 'main__As' (aka 'struct main__As')
        __chan_main__As_pushval(a, (main__Aa){EMPTY_STRUCT_INITIALIZATION});
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/test2.15413596262970579007.tmp.c:2630:71: note: passing argument to parameter 'val' here
static inline void __chan_main__As_pushval(chan_main__As ch, main__As val) {
```

--- a/vlib/v/gen/c/infix_expr.v
+++ b/vlib/v/gen/c/infix_expr.v
@@ -70,7 +70,11 @@ fn (mut g Gen) infix_expr_arrow_op(node ast.InfixExpr) {
 	}
 	g.expr(node.left)
 	g.write(', ')
-	g.expr(node.right)
+	if g.table.sym(elem_type).kind in [.sum_type, .interface_] {
+		g.expr_with_cast(node.right, node.right_type, elem_type)
+	} else {
+		g.expr(node.right)
+	}
 	g.write(')')
 	if gen_or {
 		g.or_block(tmp_opt, node.or_block, ast.void_type)

--- a/vlib/v/tests/channels_test.v
+++ b/vlib/v/tests/channels_test.v
@@ -27,3 +27,21 @@ fn test_printing_of_channels() {
 	fch.close()
 	assert fch.str() == 'chan f64{cap: 100, closed: 1}'
 }
+
+struct Aa {}
+
+struct Ab {}
+
+type As = Aa | Ab
+
+fn func(ch chan As) {
+	ch <- Aa{}
+}
+
+fn test_chan_of_sumtype() {
+	a := chan As{}
+	go func(a)
+	ret := <-a
+	println(ret)
+	assert '$ret' == 'As(Aa{})'
+}

